### PR TITLE
fix: Handle diverging ASGI `root_path` behaviour

### PR DESCRIPTION
**Pull request by provinzkraut**
As described in #2998, uvicorn change its `root_path` behaviour, which now diverges from most other ASGI servers. To ensure we handle all servers and cases properly I've implemented a fix here and added a few tests against common servers (uvicorn, hypercorn and daphne).

### CODE REVIEW

File: pyproject.toml
Comment (provinzkraut) [inline pyproject.toml:1]: Not sure why pdm reordered the file like that.. :shrug:

**Review by JacobCoffee — APPROVED** [review]

**Comment by provinzkraut** [comment]
Okay so it turns out that there was a different case where we *were* affected by this, as reported by @smithk86 in #3041. I've added a test case for this here and implemented a fix.

I've also included some basic integration tests for a few webservers with different versions for uvicorn. We should probably refactor those into their own test suite, but I didn't want to blow up this PR.

**Comment by smithk86** [comment]
@provinzkraut I tested a few combinations, and I may have found an issue. My MCVE allows for path to be `None`. On Uvicorn 0.25/Litestar 2.5.2 (with no root-path given), I can pass no `path` and get "path: None" as the response payload. On Uvicorn 0.27/Litestar-fix-uvicorn-root-path-branch though, I'm getting a 404 if no path is given.

**Comment by provinzkraut** [comment]
> @provinzkraut I tested a few combinations, and I may have found an issue. My MCVE allows for path to be `None`. On Uvicorn 0.25/Litestar 2.5.2 (**with** root-path given), I can pass no `path` and get "path: None" as the response payload. On Uvicorn 0.27/Litestar-fix-uvicorn-root-path-branch though, I'm getting a 404 if no path is given.
> 
> _edit: The above was originally wrong. Paths are handled correct without a root-path. The null path only breaks when root-path is present._

Good catch, thanks! I've added your MCVE as a test case now and made a new patch. Can you verify that everything works as expected with the newest commit?

**Comment by provinzkraut** [comment]
Tests are currently blocked by #3042

**Comment by smithk86** [comment]
> Good catch, thanks! I've added your MCVE as a test case now and made a new patch. Can you verify that everything works as expected with the newest commit?

@provinzkraut I ran though my sanity checks again, and everything looks great!

**Review by peterschutt — APPROVED** [review]
Great idea to test against the different servers!

**Comment by guacs** [comment]
@provinzkraut I had a doubt on this. Shouldn't the `root_path`, if available, be set as the prefix for the path that is created via `route_reverse` and `url_for_static_assets`? That is, which of the assertions below is the expected behavior if the server is ran using `python -m uvicorn app:app --root-path /test`?

```python
@get("/handler", name="handler")
async def handler(request: Request) -> str:
    return request.app.route_reverse("handler")

from litestar import Litestar, get, Request
from typing import List

from litestar.testing.helpers import create_test_client


@get("/handler", name="handler")
async def handler(request: Request) -> str:
    return request.app.route_reverse("handler")


with create_test_client([handler]) as client:
    response = client.get("/handler")

    url = response.json()

    assert url == "/handler"  # current behavior
    assert url == "/test/handler"  # is this the expected behavior?
```

In the [PR](https://github.com/encode/uvicorn/pull/2213) that changed the behavior in uvicorn, the following is stated:

> Uvicorn is handling an ASGI app (e.g. Starlette, FastAPI) that knows of the path /items/42, but doesn't know of the prefix /api/v1, but the app might need to create some url_for in templates or such, so it needs to also know the original path prefix it should use.

**Comment by provinzkraut** [comment]
> @provinzkraut I had a doubt on this. Shouldn't the `root_path`, if available, be set as the prefix for the path that is created via `route_reverse` and `url_for_static_assets`? 

Yes, it should be. It was an oversight, I probably should have double checked our tests for this. I’ll do a follow up PR to fix!

**Pull request merged**